sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -312,7 +312,11 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         throw new IllegalStateException(
           "logical distinct operator should have been replaced by aggregate in the optimizer")
       case logical.Repartition(numPartitions, shuffle, child) =>
-        execution.Repartition(numPartitions, shuffle, planLater(child)) :: Nil
+        if (shuffle) {
+          execution.Exchange(HashPartitioning(child.output, numPartitions), planLater(child)) :: Nil
+        } else {
+          execution.Repartition(numPartitions, shuffle, planLater(child)) :: Nil
```

Comment:
If the physical `Repartition` operator is now only used for `coalesce` then it might make sense to just rename it to `Coalesce` and to omit the `shuffle` parameter, always forcing it to `false.